### PR TITLE
importinto: check access of cloud storage uri

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #46704

Problem Summary:

### What is changed and how it works?

add check of put&delete to `br/storage`
check permission of cloud_storage_uri for importinto

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
 create a bucket on cloud without any permission, do precheck with that bucket, return access denied
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/storage/s3.go [hunk 1]
```diff
@@ -64,10 +65,11 @@ const (
...
 
-var permissionCheckFn = map[Permission]func(*s3.S3, *backuppb.S3) error{
-	AccessBuckets: checkS3Bucket,
-	ListObjects:   listObjects,
-	GetObject:     getObject,
+var permissionCheckFn = map[Permission]func(context.Context, s3iface.S3API, *backuppb.S3) error{
+	AccessBuckets:      checkS3Bucket,
+	ListObjects:        listObjects,
+	GetObject:          getObject,
+	PutAndDeleteObject: PutAndDeleteObjectCheck,
```
Comment (lance6716) [inline br/pkg/storage/s3.go:72]: naming style is slightly different 🤔  and seems no need to be public

File: br/pkg/storage/s3.go [hunk 2]
```diff
@@ -486,25 +520,30 @@ func (rs *S3Storage) IsObjectLockEnabled() bool {
 	return false
 }
 
-// WriteFile writes data to a file to storage.
-func (rs *S3Storage) WriteFile(ctx context.Context, file string, data []byte) error {
+func getPutObjectInput(options *backuppb.S3, file string, data []byte) *s3.PutObjectInput {
```
Comment (lance6716) [inline br/pkg/storage/s3.go:523]: getXXX makes me think it's a read action. Maybe buildPutObjectInput

**Review by lance6716 — COMMENTED** [review]
maybe do a manually test as well

**Comment by lance6716** [comment]
rest lgtm

**Comment by D3Hunter** [comment]
> maybe do a manually test as well

already done

File: br/pkg/storage/s3.go (hunk 1 above)
Comment (D3Hunter) [reply br/pkg/storage/s3.go:72]: > seems no need to be public

the test is in a package named as `storage_test`

File: br/pkg/storage/s3.go [hunk 3]
```diff
@@ -428,16 +430,16 @@ func NewS3Storage(ctx context.Context, backend *backuppb.S3, opts *ExternalStora
 }
 
 // checkBucket checks if a bucket exists.
-func checkS3Bucket(svc *s3.S3, qs *backuppb.S3) error {
+func checkS3BucketCheck(_ context.Context, svc s3iface.S3API, qs *backuppb.S3) error {
```
Comment (Benjamin2037) [inline br/pkg/storage/s3.go:433]: Why we need context.contex as input parameter, seems we do not use it.

Comment (tangenta) [inline br/pkg/storage/s3.go:433]:
```suggestion
func bucketExistCheck(_ context.Context, svc s3iface.S3API, qs *backuppb.S3) error {
```

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: br/pkg/storage/s3.go (hunk 3 above)
Comment (D3Hunter) [reply br/pkg/storage/s3.go:433]: new added check use it

**Review by tangenta — APPROVED** [review]

**Review by zanmato1984 — APPROVED** [review]

**Pull request merged**